example/multi_threaded_inference/multi_threaded_inference.cc

```diff
@@ -0,0 +1,351 @@
+}
+
+
+// Read Image data into a float array
+void GetImageFile(const std::string &image_file, mx_float *image_data,
+                  int channels, cv::Size resize_size) {
+  // Read all kinds of file into a BGR color 3 channels image
+  cv::Mat im_ori = cv::imread(image_file, cv::IMREAD_COLOR);
+
+  if (im_ori.empty()) {
+    std::cerr << "Can't open the image. Plase check " << image_file << ". \n";
+    assert(false);
+  }
+
+  cv::Mat im;
+  resize(im_ori, im, resize_size);
+
+  int size = im.rows * im.cols * channels;
+
+  mx_float* ptr_image_r = image_data;
+  mx_float* ptr_image_g = image_data + size / 3;
+  mx_float* ptr_image_b = image_data + size / 3 * 2;
+
+  float mean_b, mean_g, mean_r;
+  mean_b = mean_g = mean_r = DEFAULT_MEAN;
+
+  for (int i = 0; i < im.rows; ++i) {
+    auto data = im.ptr<uchar>(i);
+    for (int j = 0; j < im.cols; j++) {
+      if (channels > 1) {
+        *ptr_image_b++ = static_cast<mx_float>(*data++) - mean_b;
+        *ptr_image_g++ = static_cast<mx_float>(*data++) - mean_g;
+      }
+    }
+    *ptr_image_r++ = static_cast<mx_float>(*data++) - mean_r;
+  }
+}
+
+void prepare_input_data(const mxnet::cpp::Shape& shape, const mxnet::cpp::Context& ctx,
+                        int num_threads,
+                        std::vector<mxnet::cpp::NDArray>* data_arr,
+                        bool random_uniform = false) {
+  for (size_t i = 0; i < num_threads; ++i) {
+    data_arr->emplace_back(shape, ctx, false, 0);
+    int begin = i * 100;
+    int end = begin + 100;
+    if (random_uniform) {
+      mxnet::cpp::Operator("_random_uniform")(begin, end)
+          .Invoke((*data_arr)[i]);
+    }
+    mxnet::cpp::NDArray::WaitAll();
+  }
+}
+
+// Run inference on a model
+void run_inference(const std::string& model_name, const std::vector<mxnet::cpp::NDArray>& input_arrs,
+                   std::vector<mxnet::NDArray*> *output_mx_arr,
+                   int num_inf_per_thread = 1, bool random_sleep = false,
+                   int num_threads = 1, bool static_alloc = false,
+                   bool static_shape = false,
+                   bool is_gpu = false) {
+    LOG(INFO) << "Running inference for " + model_name +
+                 " num_threads: " + std::to_string(num_threads) +
+                 " num_inf_per_thread: " + std::to_string(num_inf_per_thread) +
+                 " random_sleep: " + std::to_string(random_sleep) +
+                 " static_alloc: " + std::to_string(static_alloc) +
+                 " static_shape: " + std::to_string(static_shape);
+  std::string json_file = model_name + "-symbol.json";
+  std::string param_file = model_name + "-0000.params";
+  auto out = mxnet::cpp::Symbol::Load(json_file);
+  std::string static_alloc_str = static_alloc ? "true" : "false";
+  std::string static_shape_str = static_shape ? "true" : "false";
+
+  // Prepare context
+# if MXNET_USE_CUDA == 1
+  mxnet::Context backend_ctx;
```

Comment:
Same question for mxnet::Context vs mxnet::cpp::Context

Reply:
frontend version of the context data structure. provides getters and setters for dev_type and dev_id.